lib/src/api/engine/tasks.rs

```diff
@@ -0,0 +1,215 @@
+use flume::Sender;
+use futures::StreamExt;
+use futures_concurrency::stream::Merge;
+#[cfg(target_arch = "wasm32")]
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+use std::time::Duration;
+
+#[cfg(not(target_arch = "wasm32"))]
+use tokio::task::JoinHandle;
+
+use surrealdb_core::dbs::Options;
+use surrealdb_core::fflags::FFLAGS;
+use surrealdb_core::kvs::Datastore;
+use surrealdb_core::options::EngineOptions;
+
+use crate::engine::IntervalStream;
+#[cfg(not(target_arch = "wasm32"))]
+use crate::Error as RootError;
+#[cfg(not(target_arch = "wasm32"))]
+use surrealdb_core::err::Error;
+#[cfg(not(target_arch = "wasm32"))]
+use tokio::spawn as spawn_future;
+#[cfg(target_arch = "wasm32")]
+use wasm_bindgen_futures::spawn_local as spawn_future;
+
+const TASK_COUNT: usize = 2;
+
+#[cfg(not(target_arch = "wasm32"))]
+type FutureTask = JoinHandle<()>;
+#[cfg(target_arch = "wasm32")]
+/// This will be true if a task has completed
+type FutureTask = Arc<AtomicBool>;
+
+/// LoggingLifecycle is used to create log messages upon creation, and log messages when it is dropped
+struct LoggingLifecycle {
+	identifier: String,
+}
+
+impl LoggingLifecycle {
+	fn new(identifier: String) -> Self {
+		debug!("Started {}", identifier);
+		Self {
+			identifier,
+		}
+	}
+}
+
+impl Drop for LoggingLifecycle {
+	fn drop(&mut self) {
+		debug!("Stopped {}", self.identifier);
+	}
+}
+
+pub struct Tasks {
+	pub nd: FutureTask,
+	pub lq: FutureTask,
+}
+
+impl Tasks {
+	#[cfg(not(target_arch = "wasm32"))]
+	pub async fn resolve(self) -> Result<(), RootError> {
+		#[cfg(not(target_arch = "wasm32"))]
+		{
+			self.nd.await.map_err(|e| {
+				error!("Node agent task failed: {}", e);
+				RootError::Db(Error::NodeAgent("node task failed and has been logged"))
+			})?;
+			self.lq.await.map_err(|e| {
+				error!("Live query task failed: {}", e);
+				RootError::Db(Error::NodeAgent("live query task failed and has been logged"))
+			})?;
+		}
+		Ok(())
+	}
+}
+
+/// Starts tasks that are required for the correct running of the engine
+pub fn start_tasks(opt: &EngineOptions, dbs: Arc<Datastore>) -> (Tasks, Vec<Sender<()>>) {
+	let mut cancellation_channels = Vec::with_capacity(TASK_COUNT);
```

Comment:
Is a heap allocation necessary here? It looks like this is always going to contain exactly two channels. Can't we use some other option that doesn't allocate on the heap like a tuple, array or a struct?